File: youtube_fix/replace.py
from __future__ import annotations

def insert_at(youtube, playlist_id: str, video_id: str, position: int | None):
    body = {
        "snippet": {
            "playlistId": playlist_id,
            "resourceId": {"kind": "youtube#video", "videoId": video_id},
        }
    }
    if position is not None:
        body["snippet"]["position"] = position
    return youtube.playlistItems().insert(part="snippet", body=body).execute()

def delete_item(youtube, playlist_item_id: str):
    return youtube.playlistItems().delete(id=playlist_item_id).execute()

File: extension/src/lib/replace.ts
// Ports insert_at and delete_item from youtube_fix/replace.py.
// Both go through apiFetch which handles token refresh.

import { apiFetch } from './youtube-api.js'
import { recordUnits } from './quota.js'

/**
 * Insert a video into a playlist at a specific 0-based position.
 * The video at that position shifts down. Returns the new playlistItemId.
 */
export async function insertAt(
  playlistId: string,
  videoId: string,
  position: number,
): Promise<string> {
  const body = {
    snippet: {
      playlistId,
      position,
      resourceId: {
        kind: 'youtube#video',
        videoId,
      },
    },
  }
  const resp = (await apiFetch('playlistItems', { part: 'snippet' }, 'POST', body)) as {
    id: string
  }
  await recordUnits(50)
  return resp.id
}

/**
 * Delete a playlist item by its playlistItemId (not by videoId or position).
 * Using piId avoids off-by-one issues after prior insertions.
 */
export async function deleteItem(piId: string): Promise<void> {
  await apiFetch('playlistItems', { id: piId }, 'DELETE')
  await recordUnits(50)
}
